notebook for certifai pro and azure hosted models

pkandarpa-cs opened the pull request:
Addresses cognitivescale/certifai#2165

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# required imports for model building and persistance \n",
```
persistance => persistence

pkandarpa-cs replied:
fixed

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create a custom prediction environment inside azure-ml workspace\n",
```
To keep this simpler and more focused on Pro, can we drop creating the custom env and just use the default one? Also reduce test cases.

pkandarpa-cs replied:
I can change this so that the following is added to pre-requisites for this notebook:

1. Service Keys for the example German Credit Model hosted on Azure ML as described in the other notebook and link them to that notebook.
2. User will only enter the Service Keys of the model here and all cells pertaining to model training and deployment can be dropped out of this notebook

cdraper-CS replied:
Good idea.

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
as discussed, please use the CLI not the internal API

pkandarpa-cs replied:
fixed

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+    "## Test the webservice\n",
+    "\n",
+    "1. Get the webservice endpoint using `service.scoring_uri` :: string\n",
+    "2. Get the authentication headers usinh `service.get_keys()` :: tuple\n"
```
usinh => using

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Running Cortex Certifai Scan locally\n",
```
This section is odd. Why's it here?

pkandarpa-cs replied:
fixed by removing

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+    "1. Fairness\n",
+    "2. Robustness\n",
+    "3. Explainability\n",
+    "4. SHAP reports"
```
SHAP reports is not an evaluation type.

pkandarpa-cs replied:
fixed

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
use CLI `certifai --version`

pkandarpa-cs replied:
fixed

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+    "\n",
+    "Additional parameters that maybe provided to the `CertifaiModel` class can be gleaned from the [API Reference for CertifaiModel](https://cognitivescale.github.io/cortex-certifai/certifai-api-ref-1.2.14/certifai.scanner.builder.html#certifai.scanner.builder.CertifaiModel)\n",
+    "\n",
+    "or `?CertifaiModel`"
```
not sure what this is doing

pkandarpa-cs replied:
It's the standard notebook doc shortcut for python objects to understand additional parameters

pkandarpa-cs replied:
an alternative to jumping to the docs site

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# run a remote scan\n",
```
remove import

pkandarpa-cs replied:
fixed

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
think things are out of order. You're getting the logs before running it

pkandarpa-cs replied:
fixed

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## View the reports from this Remote Scan\n",
```
The title is disconnected from the section, which is running the scan. There should be a section about viewing the results in the Pro console after this.

pkandarpa-cs replied:
Fixed

cdraper-CS commented on notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb:
```diff
@@ -0,0 +1,1130 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
use CLI not internal API

pkandarpa-cs replied:
fixed

pkandarpa-cs commented:
SImplified this notebook to cover the following tasks only. Other information is to. be fetched by running the original Azure Model Headers demo notebook. This notebook links to that notebook to guide users appropriately.

Please use the secrets file provided below to replace redacted secrets in the notebook for end to end testing.



[secrets.txt](https://github.com/CognitiveScale/cortex-certifai-examples/files/4686453/secrets.txt)

cdraper-CS commented:
When I load this notebook I get:
```
Notebook validation failed: 'name' is a required property:
{
 "output_type": "stream",
 "text": "\u001b[0;31mSignature:\u001b[0m \u001b[0mremote_scan\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremote_scan\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparsed_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;31mDocstring:\u001b[0m <no docstring>\n\u001b[0;31mFile:\u001b[0m      ~/miniconda3/envs/certifai-azure-model-env/lib/python3.6/site-packages/certifai/client/remote/remote_scan.py\n\u001b[0;31mType:\u001b[0m      function\n"
}
```

Also, please can you add a QA reviewer to test  before merging.

cdraper-CS commented:
I'm afraid the change to depend on the other notebook doesnt currently work. I got to the step of:
`Get the following values from the Azure ML Demo Notebook` and have no idea how to do that. I think it means I need to run through that notebook up to the point where it has deployed the models, then somehow print out those values and copy them across.   It sounded ok when we discussed it, but having tried to follow it, it seems quite cumbersome.   So maybe we need to go back to duplicating the setup :-(

cdraper-CS commented:
Can you give a bit more info for how to get the service uri and keys from the previous notebook?  I found the right place and used print(xxx) but for example there were two entries in the keys.   I _think_ it is probably workable to refer back to the previous notebook despite my initial experience where I missed the prereq statement, but its still a bit of work. 

Also would be good to refer back to the prereqs in the section where user needs the keys. For those folk like me who skipped it.

cdraper-CS commented:
Remove "Equivalent CLI commands" sections given we now only use CLI

cdraper-CS commented:
I got down to `?remote_scan.remote_scan`. This doesnt exist. There's no line that actually runs a scan. Think perhaps you are only part way through moving to CLI? Will need to write out scan definition and run `certifai remote scan...`???

cdraper-CS commented:
when I run this command `!certifai remote logs -a cpro-az -n $(certifai remote list -a cpro-az | tail -1 | cut -d' ' -f1)` I get the oldest scan not the most recent one.

cdraper-CS commented:
For viewing the reports, you may want to tell them to uncheck the "Install sample reports" when they change the storage location (assuming this actually happens). Please also give the user a bit of guidance on what they should expect to see e.g.  the use case name, that they should be able to click that to see list of scans with their scan in it, click to see results.

pkandarpa-cs commented:
@cdraper-CS I've addressed all comments with the latest update

> Can you give a bit more info for how to get the service uri and keys from the previous notebook? I found the right place and used print(xxx) but for example there were two entries in the keys. I _think_ it is probably workable to refer back to the previous notebook despite my initial experience where I missed the prereq statement, but its still a bit of work.
> 
> Also would be good to refer back to the prereqs in the section where user needs the keys. For those folk like me who skipped it.

I've added a script that generates the service keys and URIs given an Azure ML Workspace config.json

> Remove "Equivalent CLI commands" sections given we now only use CLI

Done




> I got down to `?remote_scan.remote_scan`. This doesnt exist. There's no line that actually runs a scan. Think perhaps you are only part way through moving to CLI? Will need to write out scan definition and run `certifai remote scan...`???
> 
> OK... problem is (1) the above not existing and (2) The "View Reports" section title is in wrong place. The "run scan" cell is within it

Took care of this



> when I run this command `!certifai remote logs -a cpro-az -n $(certifai remote list -a cpro-az | tail -1 | cut -d' ' -f1)` I get the oldest scan not the most recent one.
> 
> This might work better: `!certifai remote logs -a cpro-az -n $(certifai remote list -a cpro-az | head -2 | tail -1 | cut -d' ' -f1)`
> 
> The good news is I did manage to run a scan. It failed but this was expected because I didnt setup blob store etc.

Added your suggested changes



> For viewing the reports, you may want to tell them to uncheck the "Install sample reports" when they change the storage location (assuming this actually happens when you change storage location vs just on first config). Please also give the user a bit of guidance on what they should expect to see e.g. the use case name, that they should be able to click that to see list of scans with their scan in it, click to see results. Particularly important if they may have installed the samples into that location

Added information about configuring the Storage correctly in the initial cells of this notebook.

Check the rendered notebook here: https://github.com/CognitiveScale/cortex-certifai-examples/blob/certifai-2024-pro-tutorial/notebooks/azureml_model_headers_demo/german_credit_azure_ml_certifai_pro_demo.ipynb

cdraper-CS commented:
When I ran the azure_ml_service_keys.py script, I needed to `pip install ipdb`

cdraper-CS commented:
I think the names in the cleanup section may be wrong now that the user runs a script to create them.

cdraper-CS commented:
This text:
```
Instructions on how one can create and host sci-kit learn models on Azure ML for the German Credit Dataset is described in detail in the Azure ML Certifai Scan Notebook. Follow the instructions from that notebook until you have hosted trained SVM and Logistic Regression models on Azure ML with corresponding Web Service Endpoints. You'll need two items from there:
```

caused me to go and run the other tutorial. I found the instructions about running the script after I'd done that. I suggest starting out by telling them to run the script to create the hosted trained models, and THEN telling them the scripts do basically whats in the other notebook, which they can consult for more info.

pkandarpa-cs commented:
> When I ran the azure_ml_service_keys.py script, I needed to `pip install ipdb`

Got rid of it



> I think the names in the cleanup section may be wrong now that the user runs a script to create them.

Added a note in two places -> When the script is run and in the resource cleanup section about the names



> This text:
> 
> "Instructions on how one can create and host sci-kit learn models on Azure ML for the German Credit Dataset is described in detail in the Azure ML Certifai Scan Notebook. Follow the instructions from that notebook until you have hosted trained SVM and Logistic Regression models on Azure ML with corresponding Web Service Endpoints. You'll need two items from there:"
> 
> caused me to go and run the other tutorial. I found the instructions about running the script after I'd done that. I suggest starting out by telling them to run the script to create the hosted trained models, and THEN telling them the scripts do basically whats in the other notebook, which they can consult for more info.

Added this

pkandarpa-cs commented:
@bjesudasan-CS can you run through this example notebook?

pkandarpa-cs commented:
There seems to be a bug with the current deployment of ACI web services from the `azure_ml_service_keys.py` script. The deployment is unable to find the model. Fixing now

bjesudasan-CS commented:
> There seems to be a bug with the current deployment of ACI web services from the `azure_ml_service_keys.py` script. The deployment is unable to find the model. Fixing now

@pkandarpa-cs Does this address the Timeout issue?

pkandarpa-cs commented:
That’s correct. I’m testing the fix rn

On Mon, 8 Jun 2020 at 10:55, bjesudasan-CS <notifications@github.com> wrote:

> There seems to be a bug with the current deployment of ACI web services
> from the azure_ml_service_keys.py script. The deployment is unable to
> find the model. Fixing now
>
> @pkandarpa-cs <https://github.com/pkandarpa-cs> Does this address the
> Timeout issue?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/CognitiveScale/cortex-certifai-examples/pull/6#issuecomment-640376114>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGNGMHSEURNNFOQNGKPQEGLRVRY4DANCNFSM4NJ7WHRA>
> .
>

pkandarpa-cs commented:
@bjesudasan-CS you can try now! The fixed issue was:

1. scoring scripts hardcoded model name while I was using date dependent model names. This made all ACI web service deployments fail because of a missing model error. We now set environment variables pointing to the auto generated model paths in the `InferenceConfig` (Helps prevent conflicts when multiple people run these scripts on the same subscription)

pkandarpa-cs commented:
To test:

Run the script as shown above and wait until the Services are all deployed and the following output shows up:

```
❯ python scripts/azure_ml_service_keys.py --az_config ./config.json
Namespace(az_config='./config.json')
Building svm model
Model 'german_credit_svm_2020-06-08' accuracy is 0.77
Building logistic model
Model 'german_credit_logit_2020-06-08' accuracy is 0.76
model saved on disk {'model': GridSearchCV(cv=3, error_score='raise-deprecating',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'kernel': ('linear', 'rbf', 'poly'), 'C': [0.1, 0.5, 1, 2, 4, 10], 'gamma': ['auto']},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0), 'encoder': <scripts.cat_encoder.CatEncoder object at 0x10a64b0b8>}
persisted svm model to disk at german_credit_svm_2020-06-08.pkl
model saved on disk {'model': GridSearchCV(cv=3, error_score='raise-deprecating',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'C': (0.5, 1.0, 2.0), 'solver': ['lbfgs'], 'max_iter': [1000]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0), 'encoder': <scripts.cat_encoder.CatEncoder object at 0x10a64b0b8>}
persisted logistic model to disk at german_credit_logit_2020-06-08.pkl
Registering Logistic model to Azure with name german_credit_logit_2020-06-08
Registering model german_credit_logit_2020-06-08
Registering SVM model to Azure with name german_credit_svm_2020-06-08
Registering model german_credit_svm_2020-06-08
Creating Azure ML Environment with the following content
name: project_environment
dependencies:
  - python=3.6.2
  - scikit-learn==0.20.3
  - numpy==1.16.2
  - pip:
    - azureml-defaults
    - inference-schema[numpy-support]
    
Registering Webservice for german_credit_logit_2020-06-08 with name ws-tes-logis-svc-2020-06-08
Registering Webservice for german_credit_svm_2020-06-08 with name ws-tes-svm-svc-2020-06-08
Running....................................
Succeeded
ACI service creation operation finished, operation "Succeeded"
SucceededACI service creation operation finished, operation "Succeeded"
The models and services created by this script will continue running until deleted. If youre following another notebook and ran this script as part of that tutorial, wait till you"ve finished the notebook before cleaning up the resources generated by this notebook
========Azure Resources to be cleaned up later=======
Azure Resource Metadata
Workspace / ML Resource Name  --> ws-test
Subscription ID               --> REDACTED
Resource Group                --> pkandarpa
SVM Resources                 --> german_credit_svm_2020-06-08 | ws-tes-svm-svc-2020-06-08
Logistic Regression Resources --> german_credit_logit_2020-06-08 | ws-tes-logis-svc-2020-06-08
You need to take the following values and input them into your Azure Certifai Pro tutorial notebook
service_logistic_uri - REDACTED
service_logistic_key -REDACTED
service_svm_uri - REDACTED
service_svm_key - REDACTED
```

pkandarpa-cs commented:
@smichalski-cs I've updated the notebook with your suggestions.

pkandarpa-cs commented:
@smichalski-cs has given the go-ahead so I'm merging this into master
